### ISSUE TITLE
HTTPS 500 bug crash with: Abort message: 'FORTIFY: FD_SET: file descriptor >= FD_SETSIZE'

### ISSUE THREAD

**Issue by rubinbasha** [square/retrofit#1767]
- [ ] Bug report.

I am developing an app for android and I am using Retrofit for networking, below are the dependencies of the versions I use:
    compile 'com.squareup.retrofit2:retrofit:2.0.0'
    compile 'com.squareup.retrofit2:converter-gson:2.0.0'
    compile 'com.squareup.okhttp3:okhttp:3.2.0'
    compile 'com.squareup.okio:okio:1.6.0'

The app communicates always via https with the backend. There are some cases when I can get http status code 500 and in those cases the app crashes with: 

A/libc: FORTIFY: FD_SET: file descriptor >= FD_SETSIZE
A/libc: Fatal signal 6 (SIGABRT), code -6 in tid 19525 (...)

From this message and the use case when it happens I understand that there is some kind of problem where file descriptors are left open.

I did test the same use case with different endpoints and the crash seems to be related with https since it happened only in those cases. I also did test with the same endpoint with http and with https and only in the case of https it did crash. In the next comment I have attached a sample activity which just reproduces the issue. In the real life scenario the request is done every second, in the sample every 200 milliseconds so that we do not have to wait too much for it to happen. The real life app does not always do the request every second , it just does it when it does not have the data needed from the webservice.

**Comment by rubinbasha** [comment]
`public class MainActivity extends AppCompatActivity {

    private TextView tvCounter;
    private Handler handler;
    private int counter = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        tvCounter = (TextView) findViewById(R.id.tv_counter);

        handler = new Handler();
        startTimer();
    }

    private void startTimer(){
        Timer timerSchedule = new Timer();
        timerSchedule.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                makeRequest();
            }
        }, 0, 200);
    }

    private void makeRequest() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://baseurl/")
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        Service service = retrofit.create(Service.class);

        Call<Object> call = service.callService("endpoint.php");
        call.enqueue(new Callback<Object>() {
            @Override
            public void onResponse(Call<Object> call, Response<Object> response) {
                Log.e("onResponse", "onResponse status code " + response.code());
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        counter ++;
                        tvCounter.setText("" + counter);
                    }
                });
            }

            @Override
            public void onFailure(Call<Object> call, Throwable t) {
                Log.e("onFailure", "onFailure");
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        counter ++;
                        tvCounter.setText("" + counter);
                    }
                });
            }
        });
    }

    public interface Service {
        @GET
        Call<Object> callService(@Url String url);

    }
}
`

**Comment by rubinbasha** [comment]
the layout of the activity has just a TextView so that it shows how many requests were made(and to not get bored while waiting to crash :D )

Kind regards and thanks for the great libraries you offer.
Rubin Basha

**Comment by JakeWharton** [comment]
That error is in native code, presumably the code that is doing the SSL for HTTPS. Retrofit has nothing to do with that behavior and certainly has no control over it.

I would recommend asking on StackOverflow since it's unclear what is actually causing the failure and what you can do to fix it.

I can say with certainty that Retrofit plays no part in it though.

**Comment by rubinbasha** [comment]
Thank you for the reply and the hints. The problem was caused by the fact that I did not provide an OkhttpClient as a shared client for all requests. I suppose i fixed my problem now by having a shared client and using that for all requests. Problem was related to native memory leak. It is really nothing related with Retrofit as a problem but i think at least this comments should be here as a hint of how to deal with https and maybe some hint in the documentation, normally if you also think this is a solution and does not just solve a problem and open the door for other problems.

`
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://baseurl/")
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build();`

"client" is an object which is kept somewhere where all requests can access it and use as a client for all requests.
